satellite/satellitedb/nodeselection.go

```diff
@@ -0,0 +1,323 @@
+}
+
+func buildConditions(ctx context.Context, criteria *overlay.NodeCriteria, excludedIDs []storj.NodeID, excludedNetworks []string, isNewNodeQuery bool) (condition, error) {
+	var conds conditions
+	if isNewNodeQuery {
+		conds.add(
+			"total_audit_count < ?",
+			criteria.AuditCount,
+		)
+	} else {
+		conds.add(
+			"total_audit_count >= ? AND total_uptime_count >= ?",
+			criteria.AuditCount, criteria.UptimeCount,
+		)
+	}
+	conds.add(`disqualified is null
+			AND suspended is null
+			AND exit_initiated_at is null`)
+
+	conds.add("type = ?", int(pb.NodeType_STORAGE))
+	conds.add("free_disk >= ?", criteria.FreeDisk)
+	conds.add("last_contact_success > ?", time.Now().Add(-criteria.OnlineWindow))
+
+	if criteria.MinimumVersion != "" {
+		v, err := version.NewSemVer(criteria.MinimumVersion)
+		if err != nil {
+			return condition{}, Error.New("invalid node selection criteria version: %v", err)
+		}
+		conds.add(
+			"(major > ? OR (major = ? AND (minor > ? OR (minor = ? AND patch >= ?)))) AND release ",
+			v.Major, v.Major, v.Minor, v.Minor, v.Patch,
+		)
+	}
+	cond := combine(conds...)
+
+	if len(excludedIDs) > 0 {
+		excludedIDsQuery := " AND id NOT IN (?" + strings.Repeat(", ?", len(excludedIDs)-1) + ") "
+		cond.addQuery(excludedIDsQuery)
+		for _, id := range excludedIDs {
+			cond.addArg(id)
+		}
+	}
+	if criteria.DistinctIP {
+		if len(excludedNetworks) > 0 {
+			excludedNetworksQuery := " AND last_net NOT IN (?" + strings.Repeat(", ?", len(excludedIDs)-1) + ") "
+			cond.addQuery(excludedNetworksQuery)
+			for _, subnet := range excludedNetworks {
+				cond.addArg(subnet)
+			}
+		}
+		cond.addQuery(" AND last_net <> ''")
+	}
+	return cond, nil
+}
+
+func assemble(partial partialQuery) query {
+	var q strings.Builder
+	var args []interface{}
+
+	if partial.limit == 0 {
+		return query{}
+	}
+
+	fmt.Fprint(&q, partial.selection)
+
+	fmt.Fprintf(&q, " WHERE %s ", partial.condition.query)
+	args = append(args, partial.condition.args...)
+
+	if partial.orderBy != "" {
+		fmt.Fprintf(&q, " ORDER BY %s, RANDOM() ", partial.orderBy)
+	} else {
+		fmt.Fprint(&q, " ORDER BY RANDOM() ")
+	}
+
+	if !partial.distinct {
+		fmt.Fprintf(&q, " LIMIT ? ")
+		args = append(args, partial.limit)
+	}
+
+	return query{
+		query: q.String(),
+		args:  args,
+	}
+}
+
+func selectRandomFrom(q query, alias string, limit int) query {
+	if limit == 0 {
+		return query{}
+	}
+	queryString := fmt.Sprintf("SELECT * FROM (%s) %s ORDER BY RANDOM() LIMIT ?", q.query, alias)
+
+	newArgs := make([]interface{}, len(q.args)+1)
+	copy(newArgs[:len(q.args)], q.args)
+	newArgs[len(q.args)] = limit
+
+	return query{
+		query: queryString,
+		args:  newArgs,
+	}
+}
+
+func isEmpty(q query) bool {
+	if q.query == "" && len(q.args) == 0 {
+		return true
+	}
+	return false
+}
+
+func union(q1, q2 query) query {
+	if isEmpty(q1) {
+		return q2
+	}
+	if isEmpty(q2) {
+		return q1
+	}
+
+	finalQuery := "(" + q1.query + ") UNION ALL (" + q2.query + ")"
```

Comment:
why `union all` versus just `union`?

Reply:
since `union` removes duplicates but `union all` does not, I thought it would be better to use `union all` for the very slight performance gain and because there shouldn't be any duplicates between them anyway. https://www.postgresqltutorial.com/postgresql-union/